### PR TITLE
Create Environment Module, getEnvironment API method

### PR DESCRIPTION
**Pull request by johnBartos**
### This PR will...
Aggregate all information about the current session's environment into a single, lazy-loaded, memoized object.

### Why is this Pull Request needed?
Our current utils are hard-coded and clumsy to work with. 

### Are there any points in the code the reviewer needs to double check?
Wait on the compatibility script to be merged before merging this

### Are there any Pull Requests open in other repos which need to be merged with this?
https://github.com/jwplayer/jwplayer-commercial/pull/3720 must be merged after this

#### Addresses Issue(s):

JW8-122

### CODE REVIEW

File: src/js/environment/browser-version.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        } else {
+            version = agent.substring(agent.indexOf('Chrome') + 7);
+        }
+    } else if (browserEnvironment.safari) {
+        // Safari sets its version after the "Version" string in the agent
+        version = agent.substring(agent.indexOf('Version') + 8);
+    } else if (browserEnvironment.firefox) {
+        version = agent.substring(agent.indexOf('Firefox') + 8);
+    } else if (browserEnvironment.edge) {
+        // Search for "Edge/" specifically, since some rare UA's contain the word "Edge"
```
Comment (egreaves) [inline src/js/environment/browser-version.js]: You mentioned searching for `"Edge/"` in the comment but you're checking `"Edge"` here. Is this intended?

**Comment by johnBartos** [comment]
OS passes but commercial will continue to fail until it's corresponding PR is merged, which has to be merged after this PR.

File: src/js/compatibility.js [hunk 2]
```diff
@@ -72,8 +89,32 @@
...
 
+            /*
+                We've removed our browser/OS inspection utils, is* (isChrome, isAndroid, etc.) and have replaced them with
+                an Environment object. This object details the environment in which the player thinks it's in. Refer to our
+                API docs for more information.
+            */
+            var environment = playerInstance.getEnvironment();
+            var utils = playerInstance.utils;
+
+            utils.isAndroidNative = environment.OS.androidNative;
```
Comment (johnBartos) [inline src/js/compatibility.js]: Need to map to a function

File: src/js/environment/os-version.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        version = /Mac OS X (10[._\d]+)/.exec(agent)[1];
+    } else if (osEnvironment.android) {
+        version = /Android ([._\d]+)/.exec(agent)[1];
+    } else if (osEnvironment.iOS) {
+        version = /OS ([._\d]+)/.exec(agent)[1];
+    }
+
+    if (version) {
+        major = parseInt(version, 10);
+        // Versions may be in the d.d.d or d_d_d format
```
Comment (johnBartos) [inline src/js/environment/os-version.js]: Split by regex

File: test/unit/browser-version-test.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import { browserVersion } from 'environment/browser-version';
+
+describe('os-version', function () {
+    it('returns the browser version from a Chrome user agent', function () {
+        const agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36';
+        const actual = browserVersion( { chrome: true }, agent);
```
Comment (johnBartos) [inline test/unit/browser-version-test.js]: Lint error

File: src/js/environment/environment.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    flash: {
+        get: memoize(BrowserUtils.isFlashSupported),
+        enumerable: true,
+    },
+    flashVersion: {
+        get: memoize(BrowserUtils.flashVersion),
+        enumerable: true
+    },
+    iframe: {
+        get: memoize(BrowserUtils.isIframe),
```
Comment (hghazzi) [inline src/js/environment/environment.js]: memoize ftw 👍

**Comment by hghazzi** [comment]
`utils/browser` is still being used in `providers-test.js`

File: src/js/compatibility.js [hunk 6]
```diff
@@ -72,8 +89,33 @@
...
+            utils.isIE = valueFn(environment.Browser.ie);
+            utils.isIETrident = function () { return environment.Browser.ie && environment.Browser.version.major >= 11; };
+            utils.isIOS = valueFn(environment.OS.iOS);
+            utils.isIPad = valueFn(environment.OS.iPad);
+            utils.isIPod = valueFn(environment.OS.iPhone);
+            utils.isMSIE = valueFn(environment.Browser.MSIE);
+            utils.isMobile = valueFn(environment.OS.mobile);
+            utils.isOSX = valueFn(environment.OS.OSX);
+            utils.isSafari = valueFn(environment.Browser.safari);
+
```
Comment (robwalch) [inline src/js/compatibility.js]: Are you going to put these back in for the time-being to give us more time to update plugins?

**Pull request merged**